### ISSUE TITLE
babelify not being applied on .require()'d file?

### ISSUE THREAD

**Issue by chielkunkels** [babel/babelify#147]
I'm having an issue with babelify and a file I am `require`ing into my bundle. It complains about JSX syntax even though I have my presets configured. I tried dropping a `console.log()` in the `Babelify` function, just to see if it gets there, but I get the error and none of the stuff I am trying to log. Removing the require makes it work, but I need the require for my code to actually work in the browser.

I've distilled this down to a very minimal repository with a gulpfile, somewhat mimicking what I'm running into: https://github.com/chielkunkels/browserify-issue

Removing this line: https://github.com/chielkunkels/browserify-issue/blob/master/gulpfile.js#L16 makes everything work just fine. In this case the require is, of course, absolutely not needed, but in my actual code I do need it.

**Comment by bcomnes** [comment]
I'm not that familiar with grunt, but I have run into numerous bugs and unexpected behaviors when require statements wander past the top of a file.  

What happens when you consolidate all your requires at the top of the file and assign them to variables that you use later in the build script?

**Comment by zertosh** [comment]
When you call `b.require('./somefile.js')` it's not treated as a top level module so it only gets the global transforms. (It seems weird that you have the same file as an entry and require)

**Comment by chielkunkels** [comment]
Well, this is for gulp. :)

The problem here isn't in the require's of the modules in the gulpfile - the `.require()` I'm referring to is a method on browserify instances which exposes the file for other bundles.

**Comment by chielkunkels** [comment]
@zertosh that's just cause I want everything in the bundle to be exposed, that's all. I was afraid it might only receive global transforms, I had tried applying babelify globally but then I run into issues with modules that specify their own `.babelrc`.

Is there any way to force a transform to be applied to a `.require`'d file without having to resort to global transforms?

**Comment by zertosh** [comment]
That's what I'm referring to. Drop the entry and require. And try `b.add('./src/client/index.js', {expose: true})`

**Comment by chielkunkels** [comment]
Oh sweet. I wasn't aware of that one yet. Learn something new every day! Cheers.

**Comment by chielkunkels** [comment]
Hm, the problem appears to persist when using `b.add('./src/client/index.js', {expose: true})` - `SyntaxError: 'import' and 'export' may appear only with 'sourceType: module'`

**Comment by zertosh** [comment]
```js

	var appBundler = browserify(bundleOpts)
		.add('./src/client/index.js', {expose: true})

```

**Comment by chielkunkels** [comment]
Ah, I'm an idiot and forgot to add the `bundleOpts` back in (was fiddling around with it. That works a charm - thanks!

**Comment by chielkunkels** [comment]
Alright, turns out that did not entirely solve my issue. It did fix the import error, but now I cannot require modules from the first bundle in my second bundle.

I've updated the repo here: https://github.com/chielkunkels/browserify-issue/blob/master/gulpfile.js to reflect this.

It looks like `b.add()` completely ignores the `expose` option, however it does seem to use `b.require()` internally, while allowing transforms: https://github.com/substack/node-browserify/blob/master/index.js#L191

Very open to suggestions - can't help but feel I'm getting close.

**Comment by jmm** [comment]
@zertosh 
> When you call `b.require('./somefile.js')` it's not treated as a top level module

Really? I didn't know that. I wish the "top level module" thing was documented better.

> `b.add('./src/client/index.js', {expose: true})`

Undocumented :(

**Comment by jmm** [comment]
Non-global transform on `b.require()`'d file works for me.

**Comment by chielkunkels** [comment]
Got that bit working as well now (not even sure why or how, but it works), but I still can't `require('react')`, I was hoping that exposing the entry would allow me to `require()` every module in that bundle.

I also tried with `exposeAll` (undocumented, https://github.com/substack/node-browserify/blob/master/index.js#L119), but to no avail. That gives me the error: `TypeError: Path must be a string. Received false`.

It looks like I may have to use factor-bundle after all, but the problem with that approach is that I cannot mark modules as `external`. I may have to investigate that route a bit more then.

Cheers for the help so far.

**Comment by jmm** [comment]
@chielkunkels `exposeAll` is bugged: https://github.com/substack/node-browserify/issues/1293. It's public API, but needs to be documented: https://github.com/substack/node-browserify/issues/1296.

> but I still can't `require('react')`, I was hoping that exposing the entry would allow me to `require()` every module in that bundle.

No, that's on a per-module basis. That's where `exposeAll` would come in (as you tried next).

To be honest with you I've always had trouble with the exposure stuff in browserify and most of the time I use my https://github.com/jmm/pathmodify plugin to do it. I'm not exactly sure what you're trying to do or what the current problem with it is though.

BTW, most of the time when you link to a file in a repo you should make sure to use the permalink (press <kbd>y</kbd>).

**Comment by zertosh** [comment]
@jmm module-deps takes a different route to figure out if something is top level when it's not entry https://github.com/substack/module-deps/blob/d564c91/index.js#L202-L211.

@chielkunkels I suggest following what @jmm is suggesting and moving this issue to browserify, since it's not a babelify bug per-se.

**Comment by jmm** [comment]
@zertosh 
> module-deps takes a different route to figure out if something is top level when it's not entry

Sorry, I'm not sure what you mean: different than what? I've seen that code before and it looks about the same as I remember it -- it's just based on the module being in `builtins` or having a `node_modules` segment in its relative path, right? There's nothing that should inherently make a `b.require()`'d file be not top-level is there? The behavior around the `builtins` thing is questionable and possibly bugged though: https://github.com/substack/module-deps/pull/65#issuecomment-70507321.

**Comment by chielkunkels** [comment]
@jmm basically what I want is what factor-bundle does, but then force some modules to be external due to lazy-loading them.

Basically browserify's static analysis sees this code and bundles it in, even when told for it to be external. At least, that's what I experienced last time, I may just have to give it another whirl. Here's the relevant code: https://github.com/chielkunkels/browserify-issue/blob/1bb097f1e551e94c6324a616e602bff8152cc446/src/client/index.js#L8-L11

Of course this code has been extremely simplified. But basically, if I could use factor bundle but force some stuff to be external, I'm golden.

(also thanks for the permalink tip - did not know about that)

@zertosh yes, I realise this is no longer babelify territory, but I've been stuck on various incarnations of getting this to work for a few works (a few hours here and there) and I'm at my wit's end cause browserify's issue tracker isn't exactly the most active either.

**Comment by jmm** [comment]
@chielkunkels 
> Basically browserify's static analysis sees this code and bundles it in, even when told for it to be external. At least, that's what I experienced last time, I may just have to give it another whirl.

Hmm, `b.external()` or `b.exclude()` should do that. All I could suggest at this point is to strip it down to a bare bones case and take it step by step until getting each fundamental issue worked out, then build it back up to your real use case.

> browserify's issue tracker isn't exactly the most active either.

I actually think browserify should adopt the same policy as Babel for the issue tracker: bug reports and feature requests only. You could try stackoverflow, although this is a bit complicated for it.

**Comment by chielkunkels** [comment]
@jmm yeah I thought so too, I probably cocked it up somewhere along the way. At some point I had tried everything, it felt like. For now I will specify the shared modules as they come up, there shouldn't be *that* many and it'll allow to progress, can always revisit this in the future.

I have tried irc and slack both, to no avail. That's why I really appreciate you guys giving me a hand. Cheers!
